File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */

const numIslands = grid => {
    let count = 0;
    const m = grid.length;
    const n = grid[0].length;

    const dfs = (row, col) => {


        if (row < 0 || row >= m || col < 0 || col >= n) return false;
                const val = grid[row][col];
        if (val === '0' || val === 'x') return false;
        grid[row][col] = 'x'
        dfs(row + 1, col);
        dfs(row - 1, col);
        dfs(row, col + 1);
        dfs(row, col - 1);
        return true;
    }

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (dfs(i, j)) count++;
        }
    }

    return count;
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        count = 0
        
        def dfs(grid, i, j):
            if i < 0 or j < 0 or i == m or j == n or grid[i][j] == '0':
                return
            grid[i][j] = '0'
            dfs(grid, i + 1, j)
            dfs(grid, i - 1, j)
            dfs(grid, i, j + 1)
            dfs(grid, i, j - 1)
            return True
        
        for i in range(m):
            for j in range(n):
                if dfs(grid, i, j):
                    count += 1
        return count